File: polyglot/DesignPatterns/Adapter/adapter.hs
-- Adapter design pattern
import Data.Char  -- toUpper

type AudioType = String
type FileName = String

-- This the 'target' interface, namely the interface our Adapter
-- will need to implement, relying on the 'adaptee' interface
class MediaPlayer a where
  play :: a -> AudioType -> FileName -> IO ()
  play _ _ _ = putStrLn "MediaPlayer::play is not implemented"

-- This is the 'adaptee' interface, of which concrete implementations
-- provide the functionality allowing us to implement the adaptor
class IAdvancedMediaPlayer a where
  playVlc :: a -> FileName -> IO()
  playVlc _ _ = putStrLn "AdvancedMediaPlayer::playVlc is not implemented"
  playMp4 :: a -> FileName -> IO()
  playMp4 _ _ = putStrLn "AdvancedMediaPlayer::playMp4 is not implemented"

-- one implementation of the adaptee interface
data VlcPlayer = VlcPlayer
instance IAdvancedMediaPlayer VlcPlayer where
  playVlc VlcPlayer filename  = putStrLn ("Playing vlc file. Name: " ++ filename)
  playMp4 VlcPlayer _         = return () -- do nothing 

-- another implementation of the adaptee interface
data Mp4Player = Mp4Player
instance IAdvancedMediaPlayer Mp4Player where
  playVlc Mp4Player _         = return ()
  playMp4 Mp4Player filename  = putStrLn ("Playing mp4 file. Name: " ++ filename)

data AdvancedMediaPlayer = MkVlc VlcPlayer | MkMp4 Mp4Player
instance IAdvancedMediaPlayer AdvancedMediaPlayer where
  playVlc (MkVlc x) = playVlc x
  playVlc (MkMp4 x) = playVlc x
  playMp4 (MkVlc x) = playMp4 x
  playMp4 (MkMp4 x) = playMp4 x

-- Creating adapter class implementing the target interface
data MediaAdapter = MediaAdapter AudioType
advancedMusicPlayer :: MediaAdapter -> AdvancedMediaPlayer
advancedMusicPlayer (MediaAdapter audioType) =
  case map toUpper audioType of
    "VLC"   -> MkVlc VlcPlayer
    "MP4"   -> MkMp4 Mp4Player
    otherwise -> error "advancedMusicPlayer: invalid audioType"

instance MediaPlayer MediaAdapter where
  play adapter audioType filename = case map toUpper audioType of
    "VLC" -> playVlc musicPlayer filename
    "MP4" -> playMp4 musicPlayer filename
    otherwise -> error "MediaAdapter::play: invalid audioType"
    where
      musicPlayer = advancedMusicPlayer adapter

-- Creating new class of MediaPlayer which uses our adaptor (optional)
data AudioPlayer = AudioPlayer
instance MediaPlayer AudioPlayer where
  play player audioType filename = let upper = map toUpper audioType in
    -- in-built support for mp3 music file
    if upper == "MP3" then 
      putStrLn ("Playing mp3 file. Name: " ++ filename)   
    else if (upper ==  "VLC") || (upper == "MP4") then
      let adapter = MediaAdapter audioType
        in play adapter audioType filename 
    else
      putStrLn ("Invalid media. " ++ audioType ++ " format not supported")

-- We can now use the AudioPlayer to play different types of audio formats
main = let audioPlayer = AudioPlayer in do
  play audioPlayer "mp3" "beyond the horizon.mp3"
  play audioPlayer "mp4" "alone.mp4"
  play audioPlayer "vlc" "far far away.vlc"
  play audioPlayer "avi" "mind me.avi"

File: polyglot/DesignPatterns/Adapter/adapter.clj
; Adapter design pattern
(ns adapter
  (:gen-class))

; This the 'target' interface, namely the interface our Adapter
; will need to implement, relying on the 'adaptee' interface
(def MediaPlayer        ; constructor
  (letfn
    ; object created from data is message passing interface
   [(this [data]
     (fn [m]
       (cond (= m :play) (play data)
             :else (println "MediaPlayer: unknown operation error"))))
    ;
    (play [data]
      (println "MediaPlayer::play is not implemented"))]
    ; returning no argument constructor
    (fn [] (this :ignored))))
           
; This is the 'adaptee' interface, of which concrete implementations
; provide the functionality allowing us to implement the adaptor
(def AdvancedMediaPlayer; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m]
         (cond (= m :playVlc) (playVlc data)
               (= m :playMp4) (playMp4 data)
               :else (println "AdvancedMediaPlayer: unknown operation error"))))
     ;
     (playVlc [data]
       (println "AdvancedMediaPlayer::playVlc is not implemented"))
     ;
     (playMp4 [data]
       (println "AdvancedMediaPlayer::playMp4 is not implemented"))]
     ; returning no argument constructor
     (fn [] (this :ignored))))

; one implementation of the adaptee interface
(def VlcPlayer          ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m]
         (cond (= m :playVlc) (playVlc data)
               (= m :playMp4) (playMp4 data)
               :else (data m))))  ; passing message to base object (= data here)
     ;
     (playVlc [data]
       (fn [filename]
         (println "Playing vlc file. Name:" filename)))
     ;
     (playMp4 [data]
       (fn [filename] :doNothing))]
    ; returning no argument constructor
    (fn [] (this (AdvancedMediaPlayer)))))  ; passing base object as data

; another implementation of the adaptee interface
(def Mp4Player          ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m]
         (cond (= m :playVlc) (playVlc data)
               (= m :playMp4) (playMp4 data)
               :else (data m))))  ; passing message to base object (= data here)
     ;
     (playMp4 [data]
       (fn [filename]
         (println "Playing mp4 file. Name:" filename)))
     ;
     (playVlc [data]
       (fn [filename] :doNothing))]
    ; returning no argument constructor
    (fn [] (this (AdvancedMediaPlayer)))))  ; passing base object as data

; Creating adapter class implementing the target interface
(def MediaAdapter       ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m]
         (cond (= m :play) (play data)
               :else ((let [[base player] data] base) 'm)))) ; passing to base
     ;
     (play [[base player]]
       (fn [audioType filename]
         (cond (= (.toUpperCase audioType) "VLC") ((player :playVlc) filename)
               (= (.toUpperCase audioType) "MP4") ((player :playMp4) filename)
               :else :doNothing)))]
    ; returning single argument constructor
    (fn [audioType]
      (cond (= (.toUpperCase audioType) "VLC") (this [(MediaPlayer) (VlcPlayer)])
            (= (.toUpperCase audioType) "MP4") (this [(MediaPlayer) (Mp4Player)])
               :else nil))))

; Creating new class of MediaPlayer which uses our adaptor (optional)
(def AudioPlayer        ; constructor
  (letfn
    ; object created from data is message passing interface
    [(this [data]
       (fn [m]
         (cond (= m :play) (play data)
               :else (data m))))  ; passing message to base object (= data)
     ;
     (play [data]
       (fn [audioType filename]
         (cond  (= (.toUpperCase audioType) "MP3")
                (println "Playing mp3 file. Name:" filename)
                (or (= (.toUpperCase audioType) "VLC")
                    (= (.toUpperCase audioType) "MP4"))
                (let [adapter (MediaAdapter audioType)]
                  ((adapter :play) audioType filename))
                :else(println "Invalid media." audioType "format not supported"))))]
    ; returning no argument constructor
    (fn [] (this (MediaPlayer)))))

; We can now use the AudioPlayer to play different types of audio formats
(defn -main []
(def player (AudioPlayer))
((player :play) "mp3" "beyond the horizon.mp3")
((player :play) "mp4" "alone.mp4")
((player :play) "vlc" "far far away.vlc")
((player :play) "avi" "mind me.avi"))

;(-main)





